keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,296 @@
+- [Templates](#templates)
+- [Parameters](#parameters)
+- [Extensions and Bases](#extensions-and-bases)
+- [Execution State](#execution-state)
+
+These combine to form an operator responsible for the deployment, upgrading, and day 2 operations of software deployed on Kubernetes. While KUDO ships with a templating system, it is not intended to advance the state of deploying software on Kubernetes, and provides facilities for integrating with other templating and deployment systems. KUDO's focus is instead on the sequencing and day 2 operations of this software, and is able to take advantage of existing templating to fulfill this need.
+
+### Definitions
+
+### Operator Organization
+
+An operator bundle is a folder that contains all of the manifests needed to create or extend a KUDO operator. In the most basic form, an operator bundle is structured in the following format:
+
+```shell
+.
+├── operator.yaml
+├── params.yaml
+└── templates
+    ├── backup.yaml
+    ├── deployment.yaml
+    ├── init.yaml
+    ├── pvc.yaml
+    ├── restore.yaml
+    └── service.yaml
+```
+
+#### operator.yaml
+
+`operator.yaml` is the base definition of an operator. It follows the following format, extracted from the MySQL example:
+
+```yaml
+version: "5.7"
+tasks:
+  deploy:
+    resources:
+      - deployment.yaml
+      - pvc.yaml
+      - service.yaml
+  init:
+    resources:
+      - init.yaml
+  pv:
+    resources:
+      - backup-pv.yaml
+  backup:
+    resources:
+      - base-job.yaml
+    patches:
+      - backup.yaml
+  restore:
+    resources:
+      - restore.yaml
+  load-data:
+    resources:
+      - job.yaml
+  clear-data:
+    resources:
+      - job.yaml
+  query:
+    resources:
+      - job.yaml
+plans:
+  deploy:
+    steps:
+      - name: deploy
+        tasks:
+          - deploy
+      - name: init
+        tasks:
+          - init
+      - name: cleanup
+        tasks:
+          - init
+        delete: true
+  backup:
+    triggers:
+      - parameter: backupFile
+    steps:
+      - name: pv
+        tasks:
+          - pv
+      - name: backup
+        tasks:
+          - backup
+      - name: cleanup
+        tasks:
+          - backup
+        delete: true
+  restore:
+    steps:
+      - name: restore
+        tasks:
+          - restore
+      - name: cleanup
+        tasks:
+          - restore
+        delete: true
+```
+
+While subsequent sections go into deeper detail, the top level keys of the operator are:
+
+- **version**: String defining the version of a given operator
+- **tasks**: A map of tasks that can be run. These are the atomic runnable unit of a KUDO operator, and are made up of a series of YAML manifests. These are defined more in detail in [Tasks](#tasks).
+- **plans**: A map of plans that can be run. These are the core partitioning unit of a KUDO operator. A plan is intended to run a single "operations task" for an operator, such as backup, restore, deployment, or update. This is defined in detail in [Plans](#plans)
```

Comment:
I like plans with the idea of partitioning unit. Trying to understand why we took out `phases` though. I can see them as some sort of redundant to plans but liked the idea of just triggering a `plan`  with all its phases it came and not an individual `phase`.

Reply:
"Phase" just feels like we have them because we want them so far. Plans run a series of steps which execute a series of tasks. Happy to add them back in, but it just seemed redundant when a phase can always be represented as steps.

Reply:
let's talk to the DC/OS SDK team about that. I think there are examples where they're needed. I personally think phases are also nice for debugging more complex workloads like HDFS, see https://github.com/mesosphere/dcos-commons/blob/master/frameworks/hdfs/src/main/dist/svc.yml#L518-L542

Reply:
For sure. HDFS came up when talking with them about that some more. There's some key differences - for example, a phase all use a single pod, whereas that's not the case in KUDO where our base unit is any Kubernetes manifest. This enables some different ways to partition how these tasks are run.

For example, journal and name nodes might work differently in Kubernetes. @kensipe and I had a lot of discussion about this - we already have the problem of not being able to pre-reserve for ensuring journal and name nodes land on the same physical host (and have the resources available to do so), but they have independent lifecycles so can't live in the same pod. As such, we can't make even the same guarantees (other than setting affinity) around HDFS that the SDK does.

A lot of that phase complexity is dealing with that reality. The rest could be represented as a sequence of steps that are serial, with PDBs controlling parallelism at the individual Kubernetes manifest (task/template) level.

Reply:
I'm still leaning toward the simplification and driving as much as possible to Kubernetes primitives, or even introducing a new CRD (WeirdHDFSPetSet) with another controller to handle these weird cases. KUDO could even deploy whatever controller needed to handle that CRD correctly and manage pods for that use case.

Reply:
Of course, I'm cool with phases, if there's a really clear need for them, but Kubernetes can't really even guarantee the semantic captured by the SDK for HDFS (or do other semantics that require offer hoarding) without us having a custom scheduler in the first place, or having a _really_ aggressive controller that's happy to boot other things off of a node. Ultimately, for HDFS, we'll need to go down a different path but would ❤️ input from @kaiwalyajoshi, @kensipe, @timcharper, @meichstedt, and friends here.

Reply:
@gerred can we possibly make them optional or would that complicate the structure? Meaning you can specify phase if you need to, if not, you just work with tasks directly...

Reply:
@alenkacz I see where you're going, we can just make phases explicit again and avoid the complication. :)